File: simple_backend.js
#!/usr/bin/env node
/**
 * Simple Node.js backend for RAG Analytics
 * This works without Python dependencies
 */

const express = require('express');
const multer = require('multer');
const csv = require('csv-parser');
const fs = require('fs');
const path = require('path');
const cors = require('cors');

const app = express();
const PORT = 8000;

// Middleware
app.use(cors({
    origin: 'http://localhost:3000',
    credentials: true
}));
app.use(express.json());

// In-memory storage
const sessions = {};
const uploadedFiles = {};

// Configure multer for file uploads
const storage = multer.diskStorage({
    destination: (req, file, cb) => {
        cb(null, 'uploads/');
    },
    filename: (req, file, cb) => {
        cb(null, Date.now() + '-' + file.originalname);
    }
});

const upload = multer({ storage: storage });

// Create uploads directory if it doesn't exist
if (!fs.existsSync('uploads')) {
    fs.mkdirSync('uploads');
}

// Routes
app.get('/health', (req, res) => {
    res.json({ 
        status: 'healthy', 
        message: 'RAG Analytics API is running',
        timestamp: new Date().toISOString()
    });
});

app.get('/sessions', (req, res) => {
    res.json({ sessions: Object.keys(sessions) });
});

app.post('/upload_excel', upload.single('file'), (req, res) => {
    try {
        if (!req.file) {
            return res.status(400).json({ error: 'No file uploaded' });
        }

        const filePath = req.file.path;
        const fileName = req.file.originalname;
        
        // Read CSV file
        const results = [];
        fs.createReadStream(filePath)
            .pipe(csv())
            .on('data', (data) => results.push(data))
            .on('end', () => {
                // Create session
                const sessionId = `session_${Date.now()}`;
                sessions[sessionId] = {
                    type: 'file',
                    filename: fileName,
                    data: results,
                    filePath: filePath
                };

                // Get basic info
                const fileInfo = {
                    filename: fileName,
                    rows: results.length,
                    columns: results.length > 0 ? Object.keys(results[0]).length : 0,
                    column_names: results.length > 0 ? Object.keys(results[0]) : [],
                    sample_data: results.slice(0, 5),
                    data_types: results.length > 0 ? 
                        Object.keys(results[0]).reduce((acc, key) => {
                            acc[key] = typeof results[0][key];
                            return acc;
                        }, {}) : {}
                };

                res.json({
                    message: 'File uploaded successfully',
                    session_id: sessionId,
                    file_info: fileInfo
                });
            })
            .on('error', (error) => {
                console.error('Error reading file:', error);
                res.status(500).json({ error: 'Error reading file' });
            });

    } catch (error) {
        console.error('Upload error:', error);
        res.status(500).json({ error: 'Upload failed' });
    }
});

app.post('/ask_question', (req, res) => {
    try {
        const { question, session_id } = req.body;

        if (!question) {
            return res.status(400).json({ error: 'Question is required' });
        }

        if (!session_id || !sessions[session_id]) {
            return res.status(404).json({ error: 'Session not found' });
        }

        const session = sessions[session_id];
        const data = session.data;

        // Simple response for demo
        const response = {
            question: question,
            answer: `Processed question: "${question}" on dataset with ${data.length} rows and ${data.length > 0 ? Object.keys(data[0]).length : 0} columns`,
            data: {
                summary: {
                    total_rows: data.length,
                    total_columns: data.length > 0 ? Object.keys(data[0]).length : 0,
                    column_names: data.length > 0 ? Object.keys(data[0]) : []
                }
            },
            visualization: {
                type: 'table',
                data: data.slice(0, 10)
            }
        };

        res.json(response);

    } catch (error) {
        console.error('Question processing error:', error);
        res.status(500).json({ error: 'Error processing question' });
    }
});

app.get('/', (req, res) => {
    res.json({ 
        message: 'RAG Analytics API', 
        version: '1.0.0',
        endpoints: [
            'GET /health',
            'GET /sessions',
            'POST /upload_excel',
            'POST /ask_question'
        ]
    });
});

// Error handling middleware
app.use((error, req, res, next) => {
    console.error('Server error:', error);
    res.status(500).json({ error: 'Internal server error' });
});

// Start server
app.listen(PORT, () => {
    console.log('🚀 Starting Simple RAG Analytics Backend');
    console.log('==========================================');
    console.log(`Backend is running at: http://localhost:${PORT}`);
    console.log(`API docs available at: http://localhost:${PORT}`);
    console.log('Press Ctrl+C to stop the server');
    console.log();
});

// Graceful shutdown
process.on('SIGINT', () => {
    console.log('\n🛑 Shutting down server...');
    process.exit(0);
});

File: simple_backend.py
#!/usr/bin/env python3
"""
Simple FastAPI backend for RAG Analytics
This is a minimal version that works without complex dependencies
"""

from fastapi import FastAPI, File, UploadFile, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import JSONResponse
import pandas as pd
import json
import os
from typing import Dict, Any
import uvicorn

app = FastAPI(title="RAG Analytics API", version="1.0.0")

# Enable CORS for React frontend
app.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost:3000"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

# In-memory storage for demo purposes
sessions = {}
uploaded_files = {}

@app.get("/health")
async def health_check():
    """Health check endpoint"""
    return {"status": "healthy", "message": "RAG Analytics API is running"}

@app.get("/sessions")
async def get_sessions():
    """Get all active sessions"""
    return {"sessions": list(sessions.keys())}

@app.post("/upload_excel")
async def upload_excel(file: UploadFile = File(...)):
    """Upload Excel/CSV file and return basic info"""
    try:
        # Read the file
        contents = await file.read()
        
        # Save file temporarily
        file_path = f"temp_{file.filename}"
        with open(file_path, "wb") as f:
            f.write(contents)
        
        # Read with pandas
        if file.filename.endswith('.csv'):
            df = pd.read_csv(file_path)
        else:
            df = pd.read_excel(file_path)
        
        # Get basic info
        file_info = {
            "filename": file.filename,
            "rows": len(df),
            "columns": len(df.columns),
            "column_names": df.columns.tolist(),
            "sample_data": df.head(5).to_dict('records'),
            "data_types": df.dtypes.astype(str).to_dict()
        }
        
        # Store in memory
        session_id = f"session_{len(sessions) + 1}"
        sessions[session_id] = {
            "type": "file",
            "filename": file.filename,
            "dataframe": df,
            "file_path": file_path
        }
        
        # Clean up temp file
        os.remove(file_path)
        
        return {
            "message": "File uploaded successfully",
            "session_id": session_id,
            "file_info": file_info
        }
        
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Error processing file: {str(e)}")

@app.post("/ask_question")
async def ask_question(question_data: Dict[str, Any]):
    """Process natural language question"""
    try:
        question = question_data.get("question", "")
        session_id = question_data.get("session_id", "")
        
        if not question:
            raise HTTPException(status_code=400, detail="Question is required")
        
        if session_id not in sessions:
            raise HTTPException(status_code=404, detail="Session not found")
        
        session = sessions[session_id]
        df = session["dataframe"]
        
        # Simple response for demo
        response = {
            "question": question,
            "answer": f"Processed question: '{question}' on dataset with {len(df)} rows and {len(df.columns)} columns",
            "data": {
                "summary": {
                    "total_rows": len(df),
                    "total_columns": len(df.columns),
                    "column_names": df.columns.tolist()
                }
            },
            "visualization": {
                "type": "table",
                "data": df.head(10).to_dict('records')
            }
        }
        
        return response
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error processing question: {str(e)}")

@app.get("/")
async def root():
    """Root endpoint"""
    return {"message": "RAG Analytics API", "version": "1.0.0"}

if __name__ == "__main__":
    print("🚀 Starting Simple RAG Analytics Backend")
    print("==========================================")
    print("Backend will be available at: http://localhost:8000")
    print("API docs will be available at: http://localhost:8000/docs")
    print("Press Ctrl+C to stop the server")
    print()
    
    uvicorn.run(app, host="0.0.0.0", port=8000, reload=True)
